### PR TITLE
Don't access types through instance variables

### PR DESCRIPTION
**Pull request by sautermi0**
As suggested by @larshp in #288, here's the first idea of that section as a pull request.

### CODE REVIEW

File: clean-abap/CleanABAP.md [hunk 1]
```diff
@@ -2106,6 +2107,42 @@ CLASS cl_my_class IMPLEMENTATION.
...
+ENDCLASS.
+CLASS lcl IMPLEMENTATION.
+ENDCLASS.
+
+INTERFACE lif.
+  TYPES blah TYPE lcl=>foo.  
+ENDINTERFACE.
+```
+
+Using the instance for the data type would be confusing, as it suggests that the type is instance specific.
```
Comment (ConjuringCoffee) [inline clean-abap/CleanABAP.md:2129]: I generally don't agree with the inclusion of this rule, but I'd like to understand your point of view.

How is using the instance for the data type confusing? Types are never instance-specific.

**Comment by ConjuringCoffee** [comment]
Here's an example of a case in which I access a type using an instance variable. I highly simplified it of course.

```
INTERFACE interface_1.
  TYPES special_type TYPE char5.

  METHODS do_special_things
    IMPORTING
      i_special_thing TYPE special_type.
ENDINTERFACE.

INTERFACE interface_2.
  METHODS get_something_else
    RETURNING
      VALUE(result) TYPE REF TO interface_1.
ENDINTERFACE.

CLASS implementation DEFINITION CREATE PUBLIC.

  PUBLIC SECTION.
    INTERFACES interface_2.
  PROTECTED SECTION.
  PRIVATE SECTION.

ENDCLASS.

CLASS implementation IMPLEMENTATION.

  METHOD interface_2~get_something_else.
    " [...]
  ENDMETHOD.

ENDCLASS.
```

Below is the code in which I use it. I want to define a constant for the importing parameter of DO_SPECIAL_THINGS. By accessing the instance variable OTHER_OBJECT I don't need to know which type the variable even has. My only concern is that OTHER_OBJECT needs this type. If I had to explicitly use INTERFACE_1, it isn't obvious how it relates to OTHER_OBJECT. I think using INTERFACE_1 directly would be _more_ confusing.

```
  DATA object TYPE REF TO interface_2.
  object = NEW implementation( ).

  DATA(other_object) = object->get_something_else( ).

  CONSTANTS my_constant TYPE other_object->special_type VALUE '12345'.
  other_object->do_special_things( my_constant ).
```

**Comment by larshp** [comment]
a similar argument can be made against https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#dont-call-static-methods-through-instance-variables ? replace `special_type` with `static_method`

**Comment by ConjuringCoffee** [comment]
> a similar argument can be made against https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#dont-call-static-methods-through-instance-variables ? replace `special_type` with `static_method`

The topics are similar, but have two key differences to me. 

- When calling a static method on an instance, it isn't clear if the method is static or not. This isn't true for TYPES because they are never instance-specific.
- Calling a static method on an instance should never have anything to do with the actual instance directly. Calling it on the instance itself is therefore confusing.

I can understand the argument for the static methods rule, but I don't think the arguments apply to TYPES.

File: clean-abap/CleanABAP.md (hunk 1 above)
Comment (sautermi0) [reply clean-abap/CleanABAP.md:2129]: The idea was pretty exactly that: Types can never be instance-specific, so why should we make it look like the type belonged to the instance?

**Comment by sautermi0** [comment]
@ConjuringCoffee I see your point of seeing the connection between the instance and the type of the constant, though... But on the other hand, the information that "other_object" implements "interface_1" is one keyboard shortcut away...

**Comment by bjoern-jueliger-sap** [comment]
I fully support this rule:

The type is a property of the class itself, not of the instance. No other language with this kind of "associated types" (e.g. "internal typedefs" in C++ classes, "associated types" for traits in Rust) allows accessing the type though instances in the manner ABAP allows here. It is very counterintuitive and runs against how scopes/namespaces for types work in any other language. One of our explicit goals in the style guide is to write code that can be easily understood and I think this particular way of accessing types through instance variables just creates mental load on the reader, especially if ABAP is not the only language they work in.

Also, the operator `->` is generally the ABAP notation to access components of a variable behind a reference. A type is not such a component, so the notation is confusing even without comparing it to other languages. In order to understand that a type is being accessed the reader needs to realize that this particular location in an ABAP statement can only use a type, which may not always be as straightforward as one might assume. Consider the following abomination:

```abap
class cl definition final.
  public section.
    types a_type type string.
endclass.

START-OF-SELECTION.
  data(instance) = new cl( ).
  data(str) = new instance->a_type( ).
```

How long does it take you to realize that `new instance->a_type( )` is not a method call in `instance`, but a type constructor? How confident are you you'd realize this always when scanning larger passages of code?

I think the rare instances where you can make an argument that this notation might be more ergonomic don't outweigh the readability impact.

**Comment by ConjuringCoffee** [comment]
> How long does it take you to realize that new instance->a_type( ) is not a method call in instance, but a type constructor?

I was not aware that it is possible to use a type constructor in this manner. I agree that referring to a type like this is confusing and shouldn't be done.

**Review by ConjuringCoffee — APPROVED** [review]

**Review by westerholz — APPROVED** [review]

**Review by N2oB6n-SAP — APPROVED** [review]
I fully agree. I have never encountered a situation where this kind of access was anything but a shortcut for typing a few characters less. In these cases I consider the necessity to retype a longer/additional interface name to serve as nice documentation/reminder. (If typing some interface name in any given context feels odd it is usually a design/abstraction smell.)

**Review by rdibbern — APPROVED** [review]

**Review by KlausHaeuptleSAP — APPROVED** [review]

**Review by jordao76 — APPROVED** [review]

**Review by cribuc — APPROVED** [review]

**Comment by ConjuringCoffee** [comment]
It would have been nice if this was merged using the pull request title "Don't access types through instance variables" instead of the commit description "New rule" 😉 

This isn't very transparent:
image

**Pull request merged**